File: bindings/python/src/yunlink/configuration_client.py
"""Synchronous generic Configuration Service client over a YunLink Runtime."""

from __future__ import annotations

import dataclasses
import queue
import time
from typing import Any, TypeVar

from .configuration import *
from .configuration_codec import decode, encode
from .v2 import Event, Family, Peer, Qos, Runtime, Target, TypeRef

T = TypeVar("T")
_PROFILE_ID = "yunlink.core"
_PROFILE_MAJOR = 2


@dataclasses.dataclass(frozen=True)
class ConfigurationEndpoint:
    peer: Peer
    session_id: int
    endpoint_uid: str


class ConfigurationClient:
    """A single-runtime consumer for request/response Configuration operations.

    The current Python runtime exposes one owned event queue. Use this client as the
    runtime's Configuration event consumer; unrelated events remain retrievable with
    :meth:`take_deferred_events` instead of being silently discarded.
    """

    def __init__(self, runtime: Runtime, timeout_s: float = 5.0) -> None:
        self._runtime = runtime
        self._timeout_s = timeout_s
        self._deferred: list[Event] = []

    def with_timeout(self, timeout_s: float) -> "ConfigurationClient":
        return ConfigurationClient(self._runtime, timeout_s)

    def take_deferred_events(self) -> tuple[Event, ...]:
        result = tuple(self._deferred)
        self._deferred.clear()
        return result

    def _request(
        self,
        endpoint: ConfigurationEndpoint,
        request_operation: int,
        response_operation: int,
        request_type: str,
        response_type: str,
        request: Any,
        response_type_class: type[T],
    ) -> T:
        handle = self._runtime.publish(
            endpoint.peer,
            endpoint.session_id,
            Family.CONFIGURATION,
            request_operation,
            Target.endpoint(endpoint.endpoint_uid),
            TypeRef(_PROFILE_ID, _PROFILE_MAJOR, request_type),
            encode(request),
            ttl_ms=min(int(self._timeout_s * 1000), 2**32 - 1),
            qos=Qos.RELIABLE_ORDERED,
        )
        deadline = time.monotonic() + self._timeout_s
        while True:
            remaining = deadline - time.monotonic()
            if remaining <= 0:
                raise TimeoutError("YunLink Configuration response timed out")
            try:
                event = self._runtime.events.get(timeout=remaining)
            except queue.Empty as error:
                raise TimeoutError("YunLink Configuration response timed out") from error
            if self._matches(event, endpoint, handle.message_id, response_operation, response_type):
                return decode(response_type_class, event.payload)
            self._deferred.append(event)

    @staticmethod
    def _matches(
        event: Event,
        endpoint: ConfigurationEndpoint,
        correlation_id: int,
        operation: int,
        type_name: str,
    ) -> bool:
        return (
            event.peer_id == endpoint.peer.peer_id
            and event.session_id == endpoint.session_id
            and event.family == Family.CONFIGURATION
            and event.correlation_id == correlation_id
            and event.operation == operation
            and event.type_ref.profile_id == _PROFILE_ID
            and event.type_ref.major == _PROFILE_MAJOR
            and event.type_ref.type_name == type_name
        )

    def list(self, endpoint: ConfigurationEndpoint) -> ConfigResourceListResponse:
        return self._request(endpoint, 1, 2, "configuration.resource_list.request", "configuration.resource_list.response", ConfigResourceListRequest(), ConfigResourceListResponse)

    def describe(self, endpoint: ConfigurationEndpoint, resource_id: str) -> ConfigResourceDescribeResponse:
        return self._request(endpoint, 3, 4, "configuration.resource_describe.request", "configuration.resource_describe.response", ConfigResourceDescribeRequest(resource_id), ConfigResourceDescribeResponse)

    def get(self, endpoint: ConfigurationEndpoint, resource_id: str) -> ConfigResourceGetResponse:
        return self.get_variant(endpoint, resource_id, "")

    def get_variant(self, endpoint: ConfigurationEndpoint, resource_id: str, variant_id: str) -> ConfigResourceGetResponse:
        return self._request(endpoint, 5, 6, "configuration.resource_get.request", "configuration.resource_get.response", ConfigResourceGetRequest(resource_id, variant_id), ConfigResourceGetResponse)

    def patch(self, endpoint: ConfigurationEndpoint, request: ConfigResourcePatchRequest) -> ConfigResourcePatchResponse:
        return self._request(endpoint, 7, 8, "configuration.resource_patch.request", "configuration.resource_patch.response", request, ConfigResourcePatchResponse)

    def apply(self, endpoint: ConfigurationEndpoint, request: ConfigResourceApplyRequest) -> ConfigResourceApplyResponse:
        return self._request(endpoint, 9, 10, "configuration.resource_apply.request", "configuration.resource_apply.response", request, ConfigResourceApplyResponse)

    def list_variants(self, endpoint: ConfigurationEndpoint, resource_id: str) -> ConfigResourceVariantListResponse:
        return self._request(endpoint, 11, 12, "configuration.variant_list.request", "configuration.variant_list.response", ConfigResourceVariantListRequest(resource_id), ConfigResourceVariantListResponse)

    def create_variant(self, endpoint: ConfigurationEndpoint, request: ConfigResourceVariantCreateRequest) -> ConfigResourceVariantCreateResponse:
        return self._request(endpoint, 13, 14, "configuration.variant_create.request", "configuration.variant_create.response", request, ConfigResourceVariantCreateResponse)

    def save_current_as_variant(self, endpoint: ConfigurationEndpoint, request: ConfigResourceVariantSaveCurrentRequest) -> ConfigResourceVariantSaveCurrentResponse:
        return self._request(endpoint, 15, 16, "configuration.variant_save_current.request", "configuration.variant_save_current.response", request, ConfigResourceVariantSaveCurrentResponse)

    def activate_variant(self, endpoint: ConfigurationEndpoint, request: ConfigResourceVariantActivateRequest) -> ConfigResourceVariantActivateResponse:
        return self._request(endpoint, 17, 18, "configuration.variant_activate.request", "configuration.variant_activate.response", request, ConfigResourceVariantActivateResponse)

    def delete_variant(self, endpoint: ConfigurationEndpoint, request: ConfigResourceVariantDeleteRequest) -> ConfigResourceVariantDeleteResponse:
        return self._request(endpoint, 19, 20, "configuration.variant_delete.request", "configuration.variant_delete.response", request, ConfigResourceVariantDeleteResponse)


def configuration(runtime: Runtime, timeout_s: float = 5.0) -> ConfigurationClient:
    """Create an owned Configuration client without extending the C ABI."""
    return ConfigurationClient(runtime, timeout_s)

File: bindings/rust/yunlink/src/configuration_client.rs
//! Async request helper for the generic YunLink Configuration Service.

use std::time::Duration;

use tokio::time::timeout;

use crate::{
    ConfigResourceApplyRequest, ConfigResourceApplyResponse, ConfigResourceDescribeRequest,
    ConfigResourceDescribeResponse, ConfigResourceGetRequest, ConfigResourceGetResponse,
    ConfigResourceListRequest, ConfigResourceListResponse, ConfigResourcePatchRequest,
    ConfigResourcePatchResponse, ConfigResourceVariantActivateRequest,
    ConfigResourceVariantActivateResponse, ConfigResourceVariantCreateRequest,
    ConfigResourceVariantCreateResponse, ConfigResourceVariantDeleteRequest,
    ConfigResourceVariantDeleteResponse, ConfigResourceVariantListRequest,
    ConfigResourceVariantListResponse, ConfigResourceVariantSaveCurrentRequest,
    ConfigResourceVariantSaveCurrentResponse, ConfigurationPayload, Error, Event, Family, Peer,
    Qos, Result, Runtime, Target, TypeRef,
};

const TYPE_PROFILE: &str = "yunlink.core";
const TYPE_MAJOR: u16 = 2;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ConfigurationEndpoint {
    pub peer: Peer,
    pub session_id: u64,
    pub endpoint_uid: String,
}

impl ConfigurationEndpoint {
    pub fn new(peer: Peer, session_id: u64, endpoint_uid: impl Into<String>) -> Self {
        Self {
            peer,
            session_id,
            endpoint_uid: endpoint_uid.into(),
        }
    }
}

pub struct ConfigurationClient<'a> {
    runtime: &'a Runtime,
    request_timeout: Duration,
}

impl Runtime {
    pub fn configuration(&self) -> ConfigurationClient<'_> {
        ConfigurationClient {
            runtime: self,
            request_timeout: Duration::from_secs(5),
        }
    }
}

impl<'a> ConfigurationClient<'a> {
    pub fn with_timeout(mut self, value: Duration) -> Self {
        self.request_timeout = value;
        self
    }

    async fn request<Req: ConfigurationPayload, Resp: ConfigurationPayload>(
        &self,
        endpoint: &ConfigurationEndpoint,
        request_operation: u8,
        response_operation: u8,
        request_type: &'static str,
        response_type: &'static str,
        request: &Req,
    ) -> Result<Resp> {
        let mut events = self.runtime.subscribe();
        let handle = self.runtime.publish(
            &endpoint.peer,
            endpoint.session_id,
            Family::Configuration,
            request_operation,
            &Target::Endpoint(vec![endpoint.endpoint_uid.clone()]),
            &TypeRef::new(TYPE_PROFILE, TYPE_MAJOR, request_type),
            &request.encode()?,
            0,
            self.request_timeout
                .as_millis()
                .try_into()
                .unwrap_or(u32::MAX),
            Qos::ReliableOrdered,
            "",
        )?;
        timeout(self.request_timeout, async {
            loop {
                match events.recv().await.map_err(|_| Error { code: 8 })? {
                    Event::Configuration(message)
                        if message.peer_id == endpoint.peer.id
                            && message.session_id == endpoint.session_id
                            && message.correlation_id == handle.message_id
                            && message.operation == response_operation
                            && message.type_ref.profile_id == TYPE_PROFILE
                            && message.type_ref.major == TYPE_MAJOR
                            && message.type_ref.type_name == response_type =>
                    {
                        return Resp::decode(&message.payload);
                    }
                    _ => {}
                }
            }
        })
        .await
        .map_err(|_| Error { code: 8 })?
    }

    pub async fn list(
        &self,
        endpoint: &ConfigurationEndpoint,
    ) -> Result<ConfigResourceListResponse> {
        self.request(
            endpoint,
            1,
            2,
            "configuration.resource_list.request",
            "configuration.resource_list.response",
            &ConfigResourceListRequest,
        )
        .await
    }
    pub async fn describe(
        &self,
        endpoint: &ConfigurationEndpoint,
        resource_id: impl Into<String>,
    ) -> Result<ConfigResourceDescribeResponse> {
        self.request(
            endpoint,
            3,
            4,
            "configuration.resource_describe.request",
            "configuration.resource_describe.response",
            &ConfigResourceDescribeRequest {
                resource_id: resource_id.into(),
            },
        )
        .await
    }
    pub async fn get(
        &self,
        endpoint: &ConfigurationEndpoint,
        resource_id: impl Into<String>,
    ) -> Result<ConfigResourceGetResponse> {
        self.get_variant(endpoint, resource_id, "").await
    }
    pub async fn get_variant(
        &self,
        endpoint: &ConfigurationEndpoint,
        resource_id: impl Into<String>,
        variant_id: impl Into<String>,
    ) -> Result<ConfigResourceGetResponse> {
        self.request(
            endpoint,
            5,
            6,
            "configuration.resource_get.request",
            "configuration.resource_get.response",
            &ConfigResourceGetRequest {
                resource_id: resource_id.into(),
                variant_id: variant_id.into(),
            },
        )
        .await
    }
    pub async fn patch(
        &self,
        endpoint: &ConfigurationEndpoint,
        request: ConfigResourcePatchRequest,
    ) -> Result<ConfigResourcePatchResponse> {
        self.request(
            endpoint,
            7,
            8,
            "configuration.resource_patch.request",
            "configuration.resource_patch.response",
            &request,
        )
        .await
    }
    pub async fn apply(
        &self,
        endpoint: &ConfigurationEndpoint,
        request: ConfigResourceApplyRequest,
    ) -> Result<ConfigResourceApplyResponse> {
        self.request(
            endpoint,
            9,
            10,
            "configuration.resource_apply.request",
            "configuration.resource_apply.response",
            &request,
        )
        .await
    }
    pub async fn list_variants(
        &self,
        endpoint: &ConfigurationEndpoint,
        resource_id: impl Into<String>,
    ) -> Result<ConfigResourceVariantListResponse> {
        self.request(
            endpoint,
            11,
            12,
            "configuration.variant_list.request",
            "configuration.variant_list.response",
            &ConfigResourceVariantListRequest {
                resource_id: resource_id.into(),
            },
        )
        .await
    }
    pub async fn create_variant(
        &self,
        endpoint: &ConfigurationEndpoint,
        request: ConfigResourceVariantCreateRequest,
    ) -> Result<ConfigResourceVariantCreateResponse> {
        self.request(
            endpoint,
            13,
            14,
            "configuration.variant_create.request",
            "configuration.variant_create.response",
            &request,
        )
        .await
    }
    pub async fn save_current_as_variant(
        &self,
        endpoint: &ConfigurationEndpoint,
        request: ConfigResourceVariantSaveCurrentRequest,
    ) -> Result<ConfigResourceVariantSaveCurrentResponse> {
        self.request(
            endpoint,
            15,
            16,
            "configuration.variant_save_current.request",
            "configuration.variant_save_current.response",
            &request,
        )
        .await
    }
    pub async fn activate_variant(
        &self,
        endpoint: &ConfigurationEndpoint,
        request: ConfigResourceVariantActivateRequest,
    ) -> Result<ConfigResourceVariantActivateResponse> {
        self.request(
            endpoint,
            17,
            18,
            "configuration.variant_activate.request",
            "configuration.variant_activate.response",
            &request,
        )
        .await
    }
    pub async fn delete_variant(
        &self,
        endpoint: &ConfigurationEndpoint,
        request: ConfigResourceVariantDeleteRequest,
    ) -> Result<ConfigResourceVariantDeleteResponse> {
        self.request(
            endpoint,
            19,
            20,
            "configuration.variant_delete.request",
            "configuration.variant_delete.response",
            &request,
        )
        .await
    }
}
